ftml/README.md

```diff
@@ -148,6 +159,14 @@ let result = ftml::parse(&log, &tokens);
 // Now we have the final AST, as well as all the issues that
 // occurred during the parsing process.
 let (tree, warnings) = result.into();
+
+// Finally, we render with our renderer. Generally this is `HtmlRender`,
+// but you could have a custom implementation here too.
+//
+// You must provide a `PageInfo` struct, which describes the page being rendered.
+// You must also provide a handle to provide various remote sources, such as
+// module content, but this is not stabilized yet.
+let html_output = HtmlRender.render(&log, &page_info, &tree);
```

Comment:
It may be helpful to provide a way of rendering without a `page_info` object. Rendering small chunks of inline text or an editor previewing an otherwise empty draft may require making up the `page_info`, and that feels like something FTML should be able to handle.

Reply:
Actually, an FTML rendering context for inline text that avoids parsing and processing things like modules would probably help too. This pops up sometimes when writing Markdown renderers, because it's helpful when you have small chunks of formatted text to render. The easiest example I can come up with is user profiles - imagine having a `tagline` property in your profile, which is a chunk of text displayed with your user avatar. If it had formatting, it seems overkill to render it with the full FTML renderer.

Reply:
I have two points:

First, it's not that much data, just some strings that you can pretty easily mock if it's an unattached draft or something.

Second, the data is needed to properly render the page, even if it's only a snippet. Consider `[[iftags]]`, what is that supposed to do if the relevant tags aren't known?

Reply:
Good points, although on the first point I will say that every JS<->Rust interaction has a performance and latency hit, even if seemingly insignificant. The editor may be firing a preview on nearly every keystroke, keep in mind.

On the second point, yeah that's tough. In an inline context, I would just assume that `[[iftags]]` isn't valid, but in a preview context _shrug_.

Reply:
You only need to build the Rust/webasm version of the `PageInfo` struct once, right? I guess the title could change but that's not exactly on the hot path, and everything else shouldn't change at all during an editor session.

Reply:
I don't think that's the case - once I build the `PageInfo` struct, I'm fairly sure once I feed it into the `html_render` function it'll get, uh, `free`ed? I'm not sure what happens to it - it probably has something to do with Rust's borrowing. I can't reuse syntax trees, as an example.

Reply:
Oh, hmm. I thought I set everything up so they can be reused, but JS must be consuming the ownership of something. I will take a look.

Reply:
Okay, I've added the method `copy()` to all the structures. This produces a second instance of the same object, which is a very cheap operation. This way you can reuse the objects you receive / create.